### PR TITLE
[IMP] media/alt modals

### PR DESCRIPTION
**Pull request by sebgeelen**

### CODE REVIEW

File: addons/web_editor/static/src/js/wysiwyg/wysiwyg.js [hunk 1]
```diff
@@ -84,13 +87,34 @@ const Wysiwyg = Widget.extend({
             setContentEditable: this.options.setContentEditable,
         });
 
+        this._configureToolbar(options);
+
+        this.$editable.on('dblclick','img,.fa', function(e){
+            const _this = $(this);
```
Comment (Goaman) [inline addons/web_editor/static/src/js/wysiwyg/wysiwyg.js:93]: I would call it `$el` as:
- `_this` is more general
- it is more consistent with the rest of the code style to use $el

File: addons/web_editor/static/src/js/wysiwyg/wysiwyg.js [hunk 2]
```diff
@@ -589,8 +675,16 @@ const Wysiwyg = Widget.extend({
  * @returns {Node} ec - end container
  * @returns {Number} eo - end offset
  */
-Wysiwyg.getRange = function (node) {
+Wysiwyg.getRange = function (node = undefined) {
```
Comment (Goaman) [inline addons/web_editor/static/src/js/wysiwyg/wysiwyg.js:678]: I don,t see the need for the argument node in this function.
Moreover even if the argument would be used, isn't it undefined by default?

File: addons/web_editor/static/src/js/wysiwyg/wysiwyg.js [hunk 3]
```diff
@@ -340,10 +364,72 @@ const Wysiwyg = Widget.extend({
...
+            // the previous selection in not relevant anymore
+            if(mediaDialog.destroyAction !== 'save') restoreSelection();
+        });
+    },
     //--------------------------------------------------------------------------
     // Private
     //--------------------------------------------------------------------------
 
+    _configureToolbar: function (options) {
+        console.log('------\nconfigure toolbar\n------')
```
Comment (Goaman) [inline addons/web_editor/static/src/js/wysiwyg/wysiwyg.js:417]: console.log?

File: addons/web_editor/static/src/js/wysiwyg/wysiwyg.js [hunk 4]
```diff
@@ -340,10 +364,72 @@ const Wysiwyg = Widget.extend({
...
+        let $toolbar = $(this.odooEditor.toolbar);
+        console.log($toolbar, options);
```
Comment (Goaman) [inline addons/web_editor/static/src/js/wysiwyg/wysiwyg.js:419]: console.log?

File: addons/web_editor/static/src/js/wysiwyg/wysiwyg.js [hunk 5]
```diff
@@ -340,10 +364,72 @@ const Wysiwyg = Widget.extend({
...
+            e.stopImmediatePropagation();
+            e.stopPropagation();
+            this.openMediaDialog({});
+        }).bind(this));
+        $toolbar.find('#media-description').click((function(e) {
+            e.preventDefault();
+            e.stopImmediatePropagation();
+            e.stopPropagation();
+            this.openAltDialog({});
+        }).bind(this));
```
Comment (Goaman) [inline addons/web_editor/static/src/js/wysiwyg/wysiwyg.js:431]: 1. You can use the same function rather than having it duplicated.
2. You don't need to bind this if you make an arrow function.

File: addons/web_editor/static/src/js/wysiwyg/wysiwyg.js [hunk 6]
```diff
@@ -340,10 +364,72 @@ const Wysiwyg = Widget.extend({
...
+     */
+    openAltDialog(params) {
+        let $baseNode = params.node ? params.node : this.lastTargetedElement;
+
+        let altDialog = new weWidgets.AltDialog(this, params, $baseNode);
+        altDialog.open();
+    },
+    /**
+     * Open the media dialog
+     * use to insert img / icons / videos
```
Comment (Goaman) [inline addons/web_editor/static/src/js/wysiwyg/wysiwyg.js:379]: To complete the documentation, I would write:
```
Open the media dialog.

Used to insert or change image, icon, document and video.
```

File: addons/web_editor/static/src/js/wysiwyg/wysiwyg.js [hunk 7]
```diff
@@ -340,10 +364,72 @@ const Wysiwyg = Widget.extend({
...
+        // we loose the current selection inside the content editable
+        // when we click on the media dialog button
+        // so we need to be able to restore the selection when the modal is closed
+        const restoreSelection = function() {
+            if(range.sc === null) return;
+            Wysiwyg.setRange(range.sc, range.so, range. ec, range.eo);
+        }
+
+        let $baseNode = params.node ? params.node : this.lastTargetedElement;
+        let mediaDialog = new weWidgets.MediaDialog(this, params, $baseNode);
```
Comment (Goaman) [inline addons/web_editor/static/src/js/wysiwyg/wysiwyg.js:392]: what about `const`?
Just for a matter of style.

File: addons/web_editor/static/src/js/wysiwyg/wysiwyg.js [hunk 8]
```diff
@@ -340,10 +364,72 @@ const Wysiwyg = Widget.extend({
...
+     * use to insert img / icons / videos
+     */
+    openMediaDialog(params) {
+        const range = Wysiwyg.getRange();
+        // we loose the current selection inside the content editable
+        // when we click on the media dialog button
+        // so we need to be able to restore the selection when the modal is closed
+        const restoreSelection = function() {
+            if(range.sc === null) return;
+            Wysiwyg.setRange(range.sc, range.so, range. ec, range.eo);
```
Comment (Goaman) [inline addons/web_editor/static/src/js/wysiwyg/wysiwyg.js:388]: There is a space that shouldn't be there: `range. ec`

File: addons/web_editor/static/src/js/wysiwyg/wysiwyg.js [hunk 9]
```diff
@@ -340,10 +364,72 @@ const Wysiwyg = Widget.extend({
     redo: function () {
         this.odooEditor.historyRedo();
     },
+    /**
+     * Open the Alt dialog
+     * dialog give access to edit <img> alt and title
```
Comment (Goaman) [inline addons/web_editor/static/src/js/wysiwyg/wysiwyg.js:369]: I would write it with the punctuation.
Also with an emty line between the first and the second as the first is the "short description".

```
Open the Alt dialog.

Dialog give access to edit <img> alt and title.
```

File: addons/web_editor/static/src/js/wysiwyg/wysiwyg.js (hunk 1 above)
Comment (Zynton) [reply addons/web_editor/static/src/js/wysiwyg/wysiwyg.js:93]: Plus it's convention to prefix with `$` when it's a jquery element.

File: addons/web_editor/static/src/js/wysiwyg/wysiwyg.js [hunk 10]
```diff
@@ -84,13 +87,34 @@ const Wysiwyg = Widget.extend({
...
+        this._configureToolbar(options);
+
+        this.$editable.on('dblclick','img,.fa', function(e){
+            const _this = $(this);
+            let params = {"node" : _this};
+            if( _this.is('.fa')) {
+                // save layouting classes from icons to not break the page if you edit an icon
+                params.htmlClass = [..._this[0].classList].filter((className) => {
+                    return !className.startsWith('fa') || faZoomClassRegex.test(className);
+                }).join(' ');
```
Comment (Zynton) [inline addons/web_editor/static/src/js/wysiwyg/wysiwyg.js:99]: Was this not needed before? If so how come?

File: addons/web_editor/static/src/js/wysiwyg/wysiwyg.js [hunk 11]
```diff
@@ -60,6 +62,7 @@ const Wysiwyg = Widget.extend({
      */
     start: async function () {
         const _super = this._super;
+        const _self = this;
```
Comment (Zynton) [inline addons/web_editor/static/src/js/wysiwyg/wysiwyg.js:65]: The convention in Odoo is to call it `self`.

File: addons/web_editor/static/src/js/wysiwyg/wysiwyg.js [hunk 12]
```diff
@@ -84,13 +87,34 @@ const Wysiwyg = Widget.extend({
...
                 wysiwyg: this,
                 selectorEditableArea: '.o_editable',
             }, options));
             this._insertSnippetMenu();
+
+            this.snippetsMenu.on('update_customize_elements', this, (e) => {
+                this.lastTargetedElement = e.target.$target
+                // only show the description button in the toolbar if the current selected snippet is an image
+                $("#media-description").toggle(this.lastTargetedElement.is('img'));
+
```
Comment (Zynton) [inline addons/web_editor/static/src/js/wysiwyg/wysiwyg.js:116]: Superfluous space (LOC! :smile:)

File: addons/web_editor/static/src/js/wysiwyg/wysiwyg.js (hunk 9 above)
Comment (Zynton) [reply addons/web_editor/static/src/js/wysiwyg/wysiwyg.js:369]: How about this instead?
```js
/**
 * Open the Alt dialog to edit an image's `alt` and `title` attributes.
 * 
 * @param {object} params
 * @param {Node} [params.node]
 */
```
I don't know what that `params` object is besides its optional `node` property but apparently it's mandatory :shrug:

File: addons/web_editor/static/src/js/wysiwyg/wysiwyg.js [hunk 13]
```diff
@@ -340,10 +364,72 @@ const Wysiwyg = Widget.extend({
...
         this.odooEditor.historyRedo();
     },
+    /**
+     * Open the Alt dialog
+     * dialog give access to edit <img> alt and title
+     */
+    openAltDialog(params) {
+        let $baseNode = params.node ? params.node : this.lastTargetedElement;
+
+        let altDialog = new weWidgets.AltDialog(this, params, $baseNode);
```
Comment (Zynton) [inline addons/web_editor/static/src/js/wysiwyg/wysiwyg.js:374]: Both of these are constants.

File: addons/web_editor/static/src/js/wysiwyg/wysiwyg.js (hunk 6 above)
Comment (Zynton) [reply addons/web_editor/static/src/js/wysiwyg/wysiwyg.js:379]: How about this instead?
```js
/**
 * Open the Media dialog to insert or replace an image, an icon, a document or a video.
 * 
 * @param {object} params
 * @param {Node} [params.node]
 */
```
I don't know what that `params` object is besides its optional `node` property but apparently it's mandatory :shrug:

File: addons/web_editor/static/src/js/wysiwyg/wysiwyg.js [hunk 14]
```diff
@@ -340,10 +364,72 @@ const Wysiwyg = Widget.extend({
...
+                $baseNode.replaceWith(element.outerHTML);
+            } else {
+                restoreSelection();
+                document.execCommand('insertHTML', false, element.outerHTML);
+            }
+        });
+        mediaDialog.on('closed', this,  function() {
+            // if the mediaDialog content has been saved
+            // the previous selection in not relevant anymore
+            if(mediaDialog.destroyAction !== 'save') restoreSelection();
```
Comment (Zynton) [inline addons/web_editor/static/src/js/wysiwyg/wysiwyg.js:409]: Missing space before the parenthesis and missing curly braces, required by the js guidelines for Odoo.

File: addons/web_editor/static/src/js/wysiwyg/wysiwyg.js [hunk 15]
```diff
@@ -340,10 +364,72 @@ const Wysiwyg = Widget.extend({
...
+            Wysiwyg.setRange(range.sc, range.so, range. ec, range.eo);
+        }
+
+        let $baseNode = params.node ? params.node : this.lastTargetedElement;
+        let mediaDialog = new weWidgets.MediaDialog(this, params, $baseNode);
+        mediaDialog.open();
+
+        mediaDialog.on('save', this, function(element) {
+            // restore saved html classes
+            if (params.htmlClass) element.className += " " + params.htmlClass;
```
Comment (Zynton) [inline addons/web_editor/static/src/js/wysiwyg/wysiwyg.js:397]: Why is this needed?

File: addons/web_editor/static/src/js/wysiwyg/wysiwyg.js [hunk 16]
```diff
@@ -340,10 +364,72 @@ const Wysiwyg = Widget.extend({
...
+    _configureToolbar: function (options) {
+        console.log('------\nconfigure toolbar\n------')
+        let $toolbar = $(this.odooEditor.toolbar);
+        console.log($toolbar, options);
+        $toolbar.find('#media-modal').click((function(e) {
+            e.preventDefault();
+            e.stopImmediatePropagation();
+            e.stopPropagation();
+            this.openMediaDialog({});
+        }).bind(this));
```
Comment (Zynton) [inline addons/web_editor/static/src/js/wysiwyg/wysiwyg.js:425]: Why not replace this binding with an arrow function?
```js
$toolbar.find('#media-modal').click(e => {
    e.preventDefault();
    e.stopImmediatePropagation();
    e.stopPropagation();
    this.openMediaDialog({});
});
```

File: addons/web_editor/static/src/js/wysiwyg/wysiwyg.js (hunk 5 above)
Comment (Zynton) [reply addons/web_editor/static/src/js/wysiwyg/wysiwyg.js:431]: I hadn't seen that you had beat me to the punch :smile:
I was also going to recommend this:
```js
$toolbar.find('#media-modal, #media-description').click(e => {
    e.preventDefault();
    e.stopImmediatePropagation();
    e.stopPropagation();
    e.id.endsWith('modal') ? this.openMediaDialog({}) : this.openAltDialog({});
});
```
Also I notice that these are called with an empty object, which makes me wonder if it wouldn't be better to just had a check for the `params` argument within these functions so it's optional.
All of this is again question of style but that doesn't mean it doesn't matter :-)

File: addons/web_editor/static/src/js/wysiwyg/wysiwyg.js [hunk 17]
```diff
@@ -589,8 +675,16 @@ const Wysiwyg = Widget.extend({
  * @returns {Node} ec - end container
  * @returns {Number} eo - end offset
  */
-Wysiwyg.getRange = function (node) {
+Wysiwyg.getRange = function (node = undefined) {
     const selection = document.getSelection();
+    if(selection.rangeCount === 0) {
```
Comment (Zynton) [inline addons/web_editor/static/src/js/wysiwyg/wysiwyg.js:680]: Missing space.

File: addons/web_editor/static/src/xml/editor.xml [hunk 18]
```diff
@@ -114,7 +114,9 @@
             </div>
 
             <div id="link" class="btn-group">
+                <div id="media-modal" title="Insert or edit media" class="fa fa-file-image-o fa-fw btn editor-ignore"></div>
```
Comment (Zynton) [inline addons/web_editor/static/src/xml/editor.xml:117]: Are you sure the media buttons should be in the same button group as the link button? If so then we should change the id of the button group. If you do that, make sure to change references to it in template extensions.

File: addons/web_editor/static/src/js/wysiwyg/wysiwyg.js (hunk 2 above)
Comment (sebgeelen) [reply addons/web_editor/static/src/js/wysiwyg/wysiwyg.js:678]: agreed, 

I was afraid it would break some Qunit tests if we remove the argument altogether. So i just made it optionnal
It was a bit lazy, I know.

File: addons/web_editor/static/src/js/wysiwyg/wysiwyg.js (hunk 3 above)
Comment (sebgeelen) [reply addons/web_editor/static/src/js/wysiwyg/wysiwyg.js:417]: oupsys

File: addons/web_editor/static/src/js/wysiwyg/wysiwyg.js (hunk 7 above)
Comment (sebgeelen) [reply addons/web_editor/static/src/js/wysiwyg/wysiwyg.js:392]: use to be variables in the dev process, then forgot to change it back :shrug:

File: addons/web_editor/static/src/js/wysiwyg/wysiwyg.js (hunk 10 above)
Comment (sebgeelen) [reply addons/web_editor/static/src/js/wysiwyg/wysiwyg.js:99]: I don't realy know how it was made in summernode.
But this was necessary in JW integration as well.

File: addons/web_editor/static/src/js/wysiwyg/wysiwyg.js (hunk 12 above)
Comment (sebgeelen) [reply addons/web_editor/static/src/js/wysiwyg/wysiwyg.js:116]: :trolleybus:

File: addons/web_editor/static/src/js/wysiwyg/wysiwyg.js (hunk 9 above)
Comment (sebgeelen) [reply addons/web_editor/static/src/js/wysiwyg/wysiwyg.js:369]: well it can sure be refactored if necessary.
It was more or less mirrored on the openMedia function.

File: addons/web_editor/static/src/js/wysiwyg/wysiwyg.js (hunk 15 above)
Comment (sebgeelen) [reply addons/web_editor/static/src/js/wysiwyg/wysiwyg.js:397]: see comment above

File: addons/web_editor/static/src/js/wysiwyg/wysiwyg.js (hunk 16 above)
Comment (sebgeelen) [reply addons/web_editor/static/src/js/wysiwyg/wysiwyg.js:425]: yeah, indeed

File: addons/web_editor/static/src/js/wysiwyg/wysiwyg.js (hunk 5 above)
Comment (sebgeelen) [reply addons/web_editor/static/src/js/wysiwyg/wysiwyg.js:431]: :+1:

File: addons/web_editor/static/src/xml/editor.xml (hunk 18 above)
Comment (sebgeelen) [reply addons/web_editor/static/src/xml/editor.xml:117]: I just checked the mockup provided from FP in the channel. 
I'm not sure where this should be located.

Anyways, it would not be a big deal to reorganize it later.

File: addons/web_editor/static/src/js/wysiwyg/wysiwyg.js (hunk 15 above)
Comment (Zynton) [reply addons/web_editor/static/src/js/wysiwyg/wysiwyg.js:397]: Don't forget the curly braces and newline.

Comment (sebgeelen) [reply addons/web_editor/static/src/js/wysiwyg/wysiwyg.js:397]: :+1:
